SixTrack/tools/astuce/astuce++.cpp

```diff
@@ -0,0 +1,708 @@
+	{
+		std::ifstream* InputFileRead = new std::ifstream(InputFileNames_itr->c_str(), std::ios::binary);
+		char cbuffer;
+		std::streambuf* sbuffer = InputFileRead->rdbuf();
+
+		while(sbuffer->sgetc() != EOF)
+		{
+			if((cbuffer = sbuffer->sbumpc()))
+			{
+				InputFileBufferStream << cbuffer;
+			}
+		}
+
+		InputFileRead->close();
+		delete InputFileRead;
+
+		InputFileNames_itr++;
+	}
+	//*LogStream << InputFileBufferStream.str();
+	InputFileBuffer = InputFileBufferStream.str();
+}
+
+void Astuce::ParseSourceFile()
+{
+//Language
+//+cd "COMDECK"
+//+ca "CALL"
+//+if if
+//.NOT. not
+//.AND. and
+//.OR. or
+//+ei end if
+//+el else if (not used in sixtrack?)
+//+dk "DECK"
+
+	//Step 1: Extract all +cd blocks into a string map
+	ExtractFlags();
+
+	//Step 2: Extract all +dk blocks into a string map
+	ExtractDecks();
+
+	//Step 3: enable/disable +if/+ei blocks
+	ExpandIF();
+
+	//Step 4: Insert +cd blocks into the code where +ca exists
+	ExpandCA();
+}
+
+void Astuce::ExtractFlags()
+{
+	size_t CurrentPosition = 0;
+	size_t LineEnd = 0;
+	size_t Next_cd = 0;
+	size_t Next_dk = 0;
+
+	while(CurrentPosition < InputFileBuffer.size())
+	{
+		if(CurrentPosition == 0)
+		{
+			CurrentPosition = InputFileBuffer.find("+cd", CurrentPosition);
+		}
+		else
+		{
+			CurrentPosition = InputFileBuffer.find("\n+cd", CurrentPosition);
+		}
+
+		if(CurrentPosition == std::string::npos)
+		{
+			break;
+		}
+
+		LineEnd = InputFileBuffer.find("\n", CurrentPosition+1);
+
+		size_t NameStart = InputFileBuffer.rfind(" ", LineEnd);
+		std::string BlockName = FixCase(InputFileBuffer.substr(NameStart+1,LineEnd-NameStart-1));
+
+		//This could end at either the next +cd or the next +dk
+		//Find where the next one of these takes place and read until there.
+		Next_cd = InputFileBuffer.find("\n+cd", LineEnd);
+		Next_dk = InputFileBuffer.find("\n+dk", LineEnd);
+		CurrentPosition = std::min(Next_cd, Next_dk);
+
+		std::string BlockEntry = InputFileBuffer.substr(LineEnd+1, CurrentPosition-LineEnd-1);
+
+		//Now split this +cd block into lines
+		std::list<LineStorage> Lines;
+		LineStorage Entry;
+
+		size_t StartPosition = 0;
+		size_t BlockPosition = 0;
+
+		while(BlockPosition < BlockEntry.size())
+		{
+			BlockPosition = BlockEntry.find("\n", StartPosition);
+			if(BlockPosition == std::string::npos)
+			{
+				BlockPosition = BlockEntry.size();
+			}
+
+			Entry.Text = BlockEntry.substr(StartPosition, BlockPosition-StartPosition);
+			Entry.Enabled = true;
+
+			Lines.push_back(Entry);
+			StartPosition = BlockPosition+1;
+		}
+		CallStorage.insert(std::make_pair(BlockName,Lines));
+	}
+
+	if(OutputFlag)
+	{
+		std::map<std::string, std::list<LineStorage> >::const_iterator it = CallStorage.begin();
+		while(it!=CallStorage.end())
+		{
+			*LogStream << it->first << " ";
+			it++;
+		}
+		*LogStream << std::endl;
+		*LogStream << "Total \"calls\": " << CallStorage.size() << "\n" << std::endl;
+	}
+}
+
+void Astuce::ExtractDecks()
```

Comment:
Would merging these codes into one be easy? I.e. pass the function the search term and a pointer to the storage? If it's a hassle then don't bother.